### PR TITLE
DO NOT MERGE - MULE-11786 - Support API for creating an isolated Application

### PR DESCRIPTION
**Pull request by pablolagreca**

### CODE REVIEW

File: embedded/embedded-impl/src/main/java/org/mule/runtime/module/embedded/impl/EmbeddedController.java [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  private File createApplicationStructure() throws IOException {
+    File applicationFolder = new File(MuleFoldersUtil.getAppsFolder(), "app");
+    applicationFolder.mkdirs();
+    File muleArtifactFolder = new File(applicationFolder, "META-INF/mule-artifact");
+    muleArtifactFolder.mkdirs();
+    File descriptorFile = new File(muleArtifactFolder, "mule-application.json");
+
+    File pomFile = new File(artifactInfo.getPomFile().getFile());
+    Model pomModel = createModelFromPom(pomFile);
+    String pomLocation = format("META-INF%smaven%s%s%s%s%spom.xml", separator, separator, pomModel.getGroupId(), separator,
```
Comment (elrodro83) [inline embedded/embedded-impl/src/main/java/org/mule/runtime/module/embedded/impl/EmbeddedController.java]: i've seen this somwhere else

File: embedded/embedded-impl/src/main/java/org/mule/runtime/module/embedded/impl/EmbeddedController.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+  private void setUpEnvironment() {
+    setProperty(MULE_HOME_DIRECTORY_PROPERTY, containerInfo.getContainerBaseFolder().getPath());
+    getDomainsFolder().mkdirs();
+    getServicesFolder().mkdirs();
+    URLClassLoader urlClassLoader = (URLClassLoader) Thread.currentThread().getContextClassLoader();
+    URL[] urLs = urlClassLoader.getURLs();
+    for (URL urL : urLs) {
+      File dependencyFile = new File(urL.getPath());
+      if (dependencyFile.getName().endsWith("zip")) {
```
Comment (elrodro83) [inline embedded/embedded-impl/src/main/java/org/mule/runtime/module/embedded/impl/EmbeddedController.java]: case insensitive

File: embedded/embedded-impl/src/main/java/org/mule/runtime/module/embedded/impl/SerializationUtils.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.embedded.impl;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+
+public class SerializationUtils {
```
Comment (elrodro83) [inline embedded/embedded-impl/src/main/java/org/mule/runtime/module/embedded/impl/SerializationUtils.java]: is there a utils for this?

File: embedded/embedded-impl/src/test/resources/http-test-dependency/mule-config.xml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+      xsi:schemaLocation="http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
+               http://www.mulesoft.org/schema/mule/httpn http://www.mulesoft.org/schema/mule/httpn/current/mule-httpn.xsd">
+
+    <httpn:listener-config name="httpListenerConfig">
+        <httpn:listener-connection host="0.0.0.0" port="${httpPort}"/>
+    </httpn:listener-config>
+
+    <flow name="main">
+        <httpn:listener path="/" config-ref="httpListenerConfig"/>
+        <set-payload value="#[mel:Thread.currentThread().getContextClassLoader().loadClass('org.mobicents.xcap.client.impl.XcapClientImpl').getName()]"/>
```
Comment (elrodro83) [inline embedded/embedded-impl/src/test/resources/http-test-dependency/mule-config.xml]: what's this?

File: extensions/http/pom.xml [hunk 5]
```diff
@@ -64,7 +64,6 @@
             <groupId>org.mule.modules</groupId>
             <artifactId>mule-module-sockets</artifactId>
             <version>${project.version}</version>
-            <scope>provided</scope>
```
Comment (elrodro83) [inline extensions/http/pom.xml]: is this related?

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/application/AppMavenClassLoaderModelLoader.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+import org.apache.maven.model.Model;
+import org.eclipse.aether.resolution.DependencyResult;
+import org.eclipse.aether.util.graph.visitor.PreorderNodeListGenerator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is responsible of returning the {@link BundleDescriptor} of a given plugin's location and also creating a
+ * {@link ClassLoaderModel} TODO(fernandezlautaro): MULE-11094 this class is the default implementation for discovering
```
Comment (elrodro83) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/application/AppMavenClassLoaderModelLoader.java]: line break

File: bom/pom.xml [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        <artifactId>mule</artifactId>
+        <groupId>org.mule</groupId>
+        <version>4.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>mule-bom</artifactId>
+    <version>4.0-SNAPSHOT</version>
+    <packaging>pom</packaging>
+    <name>Mule Runtime BOM</name>
+    <description>Mule runtime modules for BOMs</description>
```
Comment (pablolagreca) [inline bom/pom.xml]: Build of materials. This is a maven concept to group a set of artifacts with different versions but with a relationship. In our case the relationship is a particular mule runtime version which defines the set of APIs versions

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/application/ApplicationDescriptorFactory.java [hunk 8]
```diff
@@ -109,14 +258,49 @@ public ApplicationDescriptor create(File artifactFolder) throws ArtifactDescript
       classLoaderModelBuilder.exportingPackages(jarInfo.getPackages())
           .exportingResources(jarInfo.getResources());
 
-      desc.setClassLoaderModel(classLoaderModelBuilder.build());
```
Comment (pablolagreca) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/application/ApplicationDescriptorFactory.java]: ??

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/application/EmptyApplicationDescriptor.java [hunk 9]
```diff
@@ -24,8 +24,9 @@
    */
   public EmptyApplicationDescriptor(File appLocation) {
     super(appLocation.getName());
-    setConfigResources(new String[] {DEFAULT_CONFIGURATION_RESOURCE});
-    File configPathFile = new File(appLocation, DEFAULT_CONFIGURATION_RESOURCE);
+    String configLocation = "mule" + File.separator + DEFAULT_CONFIGURATION_RESOURCE;
```
Comment (pablolagreca) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/application/EmptyApplicationDescriptor.java]: Just "mule".

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/artifact/MavenUtils.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import java.io.File;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// TODO(pablo.kraan): embedded - remove class duplication
+class MavenUtils {
+
+  private static final String USER_HOME = "user.home";
+  private static final String M2_REPO = "/.m2/repository";
```
Comment (pablolagreca) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/artifact/MavenUtils.java]: Code removed

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/artifact/MavenUtils.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    String localRepositoryProperty = getProperty("localRepository");
+    if (localRepositoryProperty == null) {
+      localRepositoryProperty = userHome + M2_REPO;
+      LOGGER.debug("System property 'localRepository' not set, using Maven default location: $USER_HOME{}", M2_REPO);
+    }
+
+    LOGGER.debug("Using Maven localRepository: '{}'", localRepositoryProperty);
+    File mavenLocalRepositoryLocation = new File(localRepositoryProperty);
+    if (!mavenLocalRepositoryLocation.exists()) {
+      throw new IllegalArgumentException("Maven repository location couldn't be found, please check your configuration");
```
Comment (pablolagreca) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/artifact/MavenUtils.java]: code removed

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java [hunk 12]
```diff
@@ -84,31 +87,42 @@ public ArtifactPluginDescriptorFactory(DescriptorLoaderRepository descriptorLoad
...
+          .orElseGet(() -> {
+            Optional<Byte[]> pluginPropertiesFileContentOptional;
+            try {
+              pluginPropertiesFileContentOptional = loadFileContentFrom(pluginJarFile, PLUGIN_PROPERTIES);
+              return loadFromPluginProperties(pluginJarFile, pluginPropertiesFileContentOptional);
+            } catch (IOException e) {
+              throw new MuleRuntimeException(e);
+            }
+          });
+    } catch (ArtifactDescriptorCreateException e) {
```
Comment (pablolagreca) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java]: Why? If the excepiton thrown is already an ArtifactDescriptorCreateException instance I don't want to wrap it.

File: embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/Repository.java [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+      final DependencyResult dependencyResult =
+          system.resolveDependencies(session, dependencyRequest);
+      final PreorderNodeListGenerator nlg = new PreorderNodeListGenerator();
+      dependencyResult.getRoot().accept(nlg);
+      return nlg;
+    } catch (DependencyResolutionException e) {
+      DependencyNode node = e.getResult().getRoot();
+      logUnresolvedArtifacts(node, e);
+      throw new RuntimeException(
+                                 String.format("There was an issue solving the dependencies for the container [%s]",
```
Comment (pablolagreca) [inline embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/Repository.java]: I changed it to "problem". I won't to mention that there's a problem more than saying that there was an exception.

File: embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/Serializer.java [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+package org.mule.runtime.module.embedded.internal;
+
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+
+public class Serializer {
```
Comment (pablolagreca) [inline embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/Serializer.java]: yes, but since this is meant to be used in other java envs, we are trying to keep at a minimum the number of external dependencies this project dependes on.

File: embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/classloading/EnumerationAdapter.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Creates an {@link Enumeration} containing all the items in a {@link Collection}
+ */
+// TODO MULE-11882 - Consolidate classloading isolation
+public class EnumerationAdapter<T> implements Enumeration<T> {
```
Comment (pablolagreca) [inline embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/classloading/EnumerationAdapter.java]: Same as before. There's one in the artifact module, but we don't want the client to also depend on that jar.

File: embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/classloading/FilteringClassLoader.java [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.Enumeration;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// TODO MULE-11882 - Consolidate classloading isolation
+public class FilteringClassLoader extends ClassLoader {
+
+  protected static final Logger logger = LoggerFactory.getLogger(FilteringClassLoader.class);
+  public static final String SYSTEM_PROPERTY_PREFIX = "mule.";
```
Comment (pablolagreca) [inline embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/classloading/FilteringClassLoader.java]: It's not configurable and that's the constant.  This class already has a TODO to consolidate classloading logic in a separate, reusable module. Once that's that, this would exists only in that module.

File: embedded/embedded-impl/src/test/resources/http-test-dependency/mule-config.xml (hunk 4 above)
Comment (pablolagreca) [inline embedded/embedded-impl/src/test/resources/http-test-dependency/mule-config.xml]: That's so the listener returns that class name as string. If it's able to do that, then it means that the application where able to load the application class successfully.

File: extensions/http/pom.xml (hunk 5 above)
Comment (pablolagreca) [inline extensions/http/pom.xml]: Yes. We have fix classloading so extensions can depend on plugins with compile scope

File: bom/pom.xml (hunk 7 above)
Comment (elrodro83) [inline bom/pom.xml]: Update the description to `Mule runtime modules for Bill of Materials (BOMs)`

File: embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/api/EmbeddedContainerFactory.java [hunk 17]
```diff
@@ -28,9 +28,22 @@
...
-  static EmbeddedContainer create(String muleVersion, URL containerBaseFolder, ArtifactInfo application) {
+/**
+ * Factory for creating an embedded container.
+ */
+final public class EmbeddedContainerFactory {
+
+  private EmbeddedContainerFactory() {}
+
+  /**
+   * Creates a container with by creating artifacts based on the {@code application} configuration.
```
Comment (elrodro83) [inline embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/api/EmbeddedContainerFactory.java]: remove `with`

File: embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/Repository.java [hunk 18]
```diff
@@ -137,17 +137,17 @@ public PreorderNodeListGenerator assemblyDependenciesForArtifact(Artifact artifa
       DependencyNode node = e.getResult().getRoot();
       logUnresolvedArtifacts(node, e);
       throw new RuntimeException(
-                                 String.format("There was an issue solving the dependencies for the container [%s]",
+                                 String.format("There was an problem solving the dependencies for the container [%s]",
```
Comment (elrodro83) [inline embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/Repository.java]: `an problem` -> `a problem`

File: embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/Repository.java [hunk 19]
```diff
@@ -137,17 +137,17 @@ public PreorderNodeListGenerator assemblyDependenciesForArtifact(Artifact artifa
...
                                                artifact),
                                  e);
     } catch (DependencyCollectionException e) {
       throw new RuntimeException(
-                                 String.format("There was an issue resolving the dependency tree for the container [%s]",
+                                 String.format("There was an problem resolving the dependency tree for the container [%s]",
```
Comment (elrodro83) [inline embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/Repository.java]: `an problem` -> `a problem`

File: embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/Repository.java [hunk 20]
```diff
@@ -137,17 +137,17 @@ public PreorderNodeListGenerator assemblyDependenciesForArtifact(Artifact artifa
...
                                                artifact),
                                  e);
     } catch (ArtifactDescriptorException e) {
       throw new RuntimeException(
-                                 String.format("There was an issue resolving the artifact descriptor for the container [%s]",
+                                 String.format("There was an problem resolving the artifact descriptor for the container [%s]",
```
Comment (elrodro83) [inline embedded/embedded-api/src/main/java/org/mule/runtime/module/embedded/internal/Repository.java]: `an problem` -> `a problem`

File: embedded/embedded-impl/src/main/java/org/mule/runtime/module/embedded/impl/EmbeddedController.java [hunk 21]
```diff
@@ -169,7 +169,7 @@ private void setUpEnvironment() {
     URL[] urLs = urlClassLoader.getURLs();
     for (URL urL : urLs) {
       File dependencyFile = new File(urL.getPath());
-      if (dependencyFile.getName().endsWith("zip")) {
+      if (dependencyFile.getName().toLowerCase().endsWith("zip")) {
```
Comment (elrodro83) [inline embedded/embedded-impl/src/main/java/org/mule/runtime/module/embedded/impl/EmbeddedController.java]: In many places this kind of code is used for dependency names. Would it make sense to have a `Dependency` abstraction or something similar? Maybe not in this PR, but at least to add a TODO.

File: modules/artifact/src/test/java/org/mule/runtime/module/artifact/builder/AbstractArtifactFileBuilder.java [hunk 22]
```diff
@@ -230,5 +220,45 @@ private void buildBrokenJarFile(File tempFile) throws UncheckedIOException {
     }
   }
 
+  /**
+   * @return a collection with custom {@link ZipResource}s to add to the artifact file.
+   */
   protected abstract List<ZipResource> getCustomResources();
+
+  /**
+   * @return the descriptor loader for the artifact. May be null.
```
Comment (elrodro83) [inline modules/artifact/src/test/java/org/mule/runtime/module/artifact/builder/AbstractArtifactFileBuilder.java]: `May be null` is redundant and misleading since the return type is an `Optional`

File: modules/artifact/src/test/java/org/mule/runtime/module/artifact/builder/AbstractDependencyFileBuilder.java [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.maven.model.Dependency;
+import org.apache.maven.model.Model;
+import org.apache.maven.model.io.xpp3.MavenXpp3Writer;
+
+/**
+ * Base class for all kind of artifact that may exists in mule.
+ *
```
Comment (elrodro83) [inline modules/artifact/src/test/java/org/mule/runtime/module/artifact/builder/AbstractDependencyFileBuilder.java]: `<p>`

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/application/DeployableMavenClassLoaderModelLoader.java [hunk 24]
```diff
@@ -34,14 +35,16 @@
 
 /**
  * This class is responsible of returning the {@link BundleDescriptor} of a given plugin's location and also creating a
- * {@link ClassLoaderModel} TODO(fernandezlautaro): MULE-11094 this class is the default implementation for discovering
+ * {@link ClassLoaderModel}
+ *
```
Comment (elrodro83) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/application/DeployableMavenClassLoaderModelLoader.java]: `<p>`

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/artifact/MavenClassLoaderModelLoader.java [hunk 25]
```diff
@@ -285,102 +272,57 @@ protected boolean enabledTestDependencies() {
...
-    session.setOffline(false);
-    session.setArtifactDescriptorPolicy((session, request) -> STRICT);
-    session.setIgnoreArtifactDescriptorRepositories(true);
-    session.setWorkspaceReader(new PomWorkspaceReader(localRepositoryDirectory, artifactFile, pluginArtifact));
-    system = repositorySystem;
-  }
-
+  /**
+   * Determines the local repository location.
+   *
```
Comment (elrodro83) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/artifact/MavenClassLoaderModelLoader.java]: `<p>`

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/artifact/MavenClassLoaderModelLoader.java [hunk 26]
```diff
@@ -285,102 +272,57 @@ protected boolean enabledTestDependencies() {
...
-    session.setWorkspaceReader(new PomWorkspaceReader(localRepositoryDirectory, artifactFile, pluginArtifact));
-    system = repositorySystem;
-  }
-
+  /**
+   * Determines the local repository location.
+   * 
+   * If the artifact is being loaded from an application repository, then such repository must be used as the local maven
+   * repository.
+   *
```
Comment (elrodro83) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/artifact/MavenClassLoaderModelLoader.java]: `<p>`

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/artifact/MavenClassLoaderModelLoader.java [hunk 27]
```diff
@@ -285,102 +272,57 @@ protected boolean enabledTestDependencies() {
...
-  }
-
+  /**
+   * Determines the local repository location.
+   * 
+   * If the artifact is being loaded from an application repository, then such repository must be used as the local maven
+   * repository.
+   * 
+   * If the artifact doesn't belong to an application repository then the container repository must be used instead.
+   *
```
Comment (elrodro83) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/artifact/MavenClassLoaderModelLoader.java]: `<p>`

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/PluginMavenClassLoaderModelLoader.java [hunk 28]
```diff
@@ -33,10 +33,11 @@
 
 /**
  * This class is responsible of returning the {@link BundleDescriptor} of a given plugin's location and also creating a
- * {@link ClassLoaderModel} TODO(fernandezlautaro): MULE-11094 this class is the default implementation for discovering
- * dependencies and URLs, which happens to be Maven based. There could be other ways to look for dependencies and URLs (probably
- * for testing purposes where the plugins are done by hand and without maven) which will imply implementing the jira pointed out
- * in this comment.
+ * {@link ClassLoaderModel}
+ *
```
Comment (elrodro83) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/PluginMavenClassLoaderModelLoader.java]: `<p>`

File: modules/deployment-model-impl/src/test/java/org/mule/runtime/module/deployment/impl/internal/application/ApplicationDescriptorFactoryTestCase.java [hunk 29]
```diff
@@ -188,8 +188,8 @@ public void applicationDescriptorFromJson() throws Exception {
     ApplicationDescriptor desc = createApplicationDescriptor(appPath);
 
     assertThat(desc.getMinMuleVersion(), is(new MuleVersion("4.0.0")));
-    assertThat(desc.getConfigResources().length, is(1));
-    assertThat(desc.getConfigResources()[0], is("mule/" + DEFAULT_CONFIGURATION_RESOURCE));
+    assertThat(desc.getConfigResources().size(), is(1));
```
Comment (elrodro83) [inline modules/deployment-model-impl/src/test/java/org/mule/runtime/module/deployment/impl/internal/application/ApplicationDescriptorFactoryTestCase.java]: Use `IsCollectionWithSize` matcher

File: modules/deployment-model-impl/src/test/java/org/mule/runtime/module/deployment/impl/internal/application/ApplicationDescriptorFactoryTestCase.java [hunk 30]
```diff
@@ -242,8 +242,8 @@ public void classLoaderModelWithPluginDependencyWithAnotherPlugin() throws Excep
 
     ClassLoaderModel classLoaderModel = desc.getClassLoaderModel();
 
-    assertThat(classLoaderModel.getDependencies().size(), is(2));
-    assertThat(classLoaderModel.getDependencies(), hasItems(socketsPluginDependencyMatcher(), httpPluginDependencyMatcher()));
+    assertThat(classLoaderModel.getDependencies().size(), is(1));
```
Comment (elrodro83) [inline modules/deployment-model-impl/src/test/java/org/mule/runtime/module/deployment/impl/internal/application/ApplicationDescriptorFactoryTestCase.java]: use `IsCollectionWithSize` matcher

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java [hunk 31]
```diff
@@ -1812,6 +1826,7 @@ private void doSynchronizedAppDeploymentActionTest(final Action deploymentAction
                                                emptyAppFileBuilder.getId());
   }
 
+  @Ignore("TODO MULE-11551 - Define mechanism to inform shared libraries dependencies")
```
Comment (elrodro83) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java]: no need to put TODO in `@Ignore`

File: tests/runner/src/main/java/org/mule/test/runner/api/AetherClassPathClassifier.java [hunk 32]
```diff
@@ -357,18 +357,18 @@ public ArtifactsUrlClassification classify(final ClassPathClassifierContext cont
...
    *
    * @param rootArtifactType the {@link ArtifactClassificationType} for rootArtifact
    * @return {@link Predicate} for selecting direct dependencies for the Container.
    */
   private Predicate<Dependency> getContainerDirectDependenciesFilter(ArtifactClassificationType rootArtifactType) {
     return rootArtifactType.equals(MODULE)
         ? directDep -> directDep.getScope().equals(PROVIDED) || directDep.getScope().equals(COMPILE)
-        : directDep -> directDep.getScope().equals(PROVIDED);
+        : directDep -> directDep.getScope().equals(PROVIDED)
+            || directDep.getArtifact().getClassifier().equals(MULE_PLUGIN_CLASSIFIER);
```
Comment (gsfernandes) [inline tests/runner/src/main/java/org/mule/test/runner/api/AetherClassPathClassifier.java]: @pablolagreca, I working on this code in order to remove the constraint of having to define plugins as PROVIDED for isolated tests. I was wondering if you remember why this filter has to be changed for including mule-plugins as part of the filter for the Container URL.
